sdk/src/main/java/io/customer/sdk/queue/QueueRequestManager.kt

```diff
@@ -0,0 +1,52 @@
+package io.customer.sdk.queue
+
+// Make sure implementation is thread-safe
+interface QueueRequestManager {
+    /**
+     * Have the queue call this when it completes running. This frees up the SDK to run the request again on the next request.
+     */
+    fun queueRunRequestComplete()
+
+    /**
+     * Have the queue call this when it wants to start a new request. Manager will remember that the queue is running a request.
+     *
+     * @return if an existing run request is currently running. If true, the queue should not run the queue.
+     */
+    fun startRequest(): Boolean
+}
+
+/**
+Simple singleton (separate singleton for each siteId) that helps assert that the background queue is only running
+one run request at one time (1 run request per site id since each site id can have it's own background queue).
+When the background queue wants to run it's tasks, it's important that the queue only have 1 concurrent runner
+running at one time to prevent race conditions and tasks running multiple times.
+This class is small and separate from the rest of the queue logic for some readability/scalability value but
+mostly memory safety.
+We want to avoid making our queue classes singletons because these classes may have lots of
+dependencies inside of them (especially the runner). We want to avoid keeping all of these dependencies sitting in
+memory.
+ */
+class QueueRequestManagerImpl : QueueRequestManager {
+
+    @Volatile var isRunningRequest: Boolean = false
```

Comment:
Maybe this is premature, but I don't understand the concept of `QueueRequestManager` because this class is just holding a variable that we are using as a singleton.

We can just have `isRunningRequest` volatile variable in the `QueueRunRequest`, whats the benefit of having it in a separate class?

Reply:
This was a popular question with the iOS implementation, too! Great question, too. I would ask the same thing. 

I tried to explain the reason for `QueueRequestManager` in the comments at the top of the file. Specifically:
```
...mostly memory safety.

We want to avoid making our queue classes singletons because these classes may have lots of
dependencies inside of them (especially the runner). We want to avoid keeping all of these dependencies sitting in
memory.
```

Let me try to explain it better then I can update the comment in `QueueRequestManager`. 

Problem: We need to make sure that there is only 1 background queue (per site-id) that is running. Therefore, we need to keep track of the state of the background queue if it is running or if it is not. Keeping track of the state probably requires the use of a singleton. 

The only class that uses `QueueRequestManager` is `QueueRunRequest` as it's in charge of running queue tasks. We might ask, could we just make `QueueRunRequest` a singleton and move all logic from `QueueRequestManager` into it? 

I wanted to avoid doing that because `QueueRunRequest` has a dependency on `QueueRunner` and `QueueRunner` has the potential of having *lots* of dependencies in it. The dependency tree that begin to form from the starting node of `QueueRunRequest` can grow quite big and that tree would stay in memory because it belongs to a singleton. 

So, by encapsulating the singleton that we need into as small of a class as possible, we avoid this issue. 

Rethinking this, what might work is to refactor `QueueRunRequest` to be a singleton but manage all of it's own dependencies inside. For example, after `QueueRunner` is no longer being used in the class, the `QueueRunRequest` would set that property to `null` and have it GCed. Maybe that would work? Any other ideas?

Reply:
Your concern is very legit, I just want to check on a few things. 

> We need to make sure that there is only 1 background queue (per site-id) that is running

So from the code, an instance of `Queue` is created when an singleton instance of `CustomerIO` is created, in case a new `site-id` is needed, a new instance of `CustomerIO` and hence a new instance of `Queue` will be created. 

Inside `Queue` we have `QueueRunRequest`, so my assumption is if we aren't going to have another instance of `Queue` in a session, we can't have another instance of `QueueRunRequest` either as well. Do you agree with that assumption? 

Now that we know, we can't multiple instances of `QueueRunRequest`, what I think might even work is instead of having `QueueRequestManager` we can just have 

`@Volatile var isRunningRequest: Boolean = false` and all of its supporting methods inside, `QueueRunRequest` if we to make sure, one queue is running at a time. 

Otherwise, we can opt with what you are proposing as well.

Reply:
> so my assumption is if we aren't going to have another instance of Queue in a session, we can't have another instance of QueueRunRequest either as well. Do you agree with that assumption?

In the code as it is in the PR right now, `Queue` is *not* a singleton. You could have 2 repositories exist in memory, each with it's own `Queue` instance, as an example. If `Queue` was a singleton then yes, your comment after "Now that we know," would be correct and we would not need `QueueRequestManager`.

Reply:
> I believe this can be achieved even now.

It would be great if we can *be sure* that this can be achieved now. 

Something I want us to consider for this SDK is to implement something like what I recently did for the [iOS SDK](https://github.com/customerio/customerio-ios/pull/137). I created a file [for each module](https://github.com/customerio/customerio-ios/blob/0ca06183f9c0094019e3e58d0963b5931909c348/Tests/Tracking/APITest.swift) that contains code that calls all of the public SDK functions. 

This does a few things for us. (1) asserts that we remember to make public functions `public` access, (2) assert customers are able to mock every function, (3) asserts we do not accidentally make any breaking changes to the SDK.

Reply:
Change implemented in this commit: https://github.com/customerio/customerio-android/pull/67/commits/f44c91e60e6c0e21c00466dee6a30b426c794854